### DISCUSSION TITLE
Idées en vue d'une refactorisation

### DISCUSSION THREAD

**Discussion by elias75015** [IGNF/geopf-extensions-openlayers#275] [Ideas]
Dans ce ticket chapeau, l'idée dans un premier temps est de lister ce qui nous parait améliorable dans le code pour permettre une homogénéisation de la manière de créer un widget.

Dans un second temps, on essayera de décliner les différentes remontées listées ici en sous-tickets au périmètre mieux défini.

A terme, l'objectif est de proposer un protocole de contribution basé sur un workflow bien documenté et éventuellement des templates.

### Idées Générales

- Créer un "Core" pour la création du DOM : une bibliothèque de fonctions dans laquelle un contributeur pourrait piocher pour créer son widget.

- Faire la distinction entre plusieurs types de widgets : widget avec formulaire (ex.import de donnéees), widget avec interaction (ex.calcul de distance)

- Homogénéiser les widgets fullCustom et les widgets natifs OpenLayers qu'on surcharge

### Incohérences entre widgets

- les options du widgets sont passées à des endroits différents

- des options similaires entre deux widgets ont des nom différents

- la manière d'ajouter des classCSS aux éléments (parfois classList.add(), parfois classList = ".....")

### Factorisations possibles

- le constructeur des widgets est souvent surchargé avec les mêmes éléments

- les setters/getter du DOM (ex. setCollapsed, getContainer)

- pour les widgets de calcul/dessin, les méthodes pour récupérer la données (set/get data, set/getgeometry)

- la création du DOM et des listeners sur les boutons (ex. _createShowIsoExclusionsPictoElement = _createShowElevationPathPictoElement = ....)

**Comment by IGNFhc** [comment]
- Uniformiser et rendre paramétrable les styles (carto et DOM)
- ajouter des méthodes de DOM getter et setter (icone, panel, ....)
- templatiser le DOM
- avoir un modèle de données plus précis (utilisation du Typescript, standardisation des entrées et sortie d’un widget)
- centraliser certaines infos : liste et ordre des couches, objet/layer en interaction sur la carte, interaction entre les widgets (collision, ...)
par exemple : des méthodes pour fermer tous les widget de la carte
- générateur de widget : quand on veut créer un widget on peut faire npm create widget, ça pose qulques questions et bim. Ya plus qu’à coder fonctionnellement et spécifiquement ce qui s’y passe.
(un widget à besoin: d’infos de la map / de certaines interactions avec la map / un code spécifique / du dom à peupler en fonction de ces infos)

**Comment by elias75015** [comment]
Link : https://github.com/IGNF/geopf-extensions-openlayers/issues/39

**Comment by azarz** [comment]
Centraliser un maximum de code commun à tous les widgets héritant de ControlExtended dans cette dernière :
Création de l'uuid (https://github.com/IGNF/geopf-extensions-openlayers/pull/284), getContainer(), etc

**Comment by elias75015** [comment]
@IGNF/geopf-api 

Première réunion du 08/04/2025, liste des tâches pour la prochaine échéance : 

-1) faire diagramme UML de l'existant --> Elias 
-2) explorer autres API/produits qui existent --> Hector, Amaury, JP
-3) faire nouveau UML -> Hector
-4) lister incohérences syntaxiques -> Elias
-5) définir linter pour refacetorisation -> Elias, Hector

Prochain point le 24/04

**Reply by lowzonenose** [reply]
Perso, j'aime bien comment est réalisé cette lib UI, elle répond à pas mal de nos besoins
https://github.com/Polydile/dile-components

**Comment by elias75015** [comment]
![plant_uml_ext_geopf](https://github.com/user-attachments/assets/dc4cb4e0-951e-4768-b224-00ae6e5c1a2d)
Tache 1 : Premier jet à compléter corriger

**Reply by elias75015** [reply]
[geopf_widgets_full_uml.txt](https://github.com/user-attachments/files/19762610/geopf_widgets_full_uml.txt)

Fichier plant uml en .puml

**Reply by elias75015** [reply]
[geopf_widgets_full_extended_uml.txt](https://github.com/user-attachments/files/19812620/geopf_widgets_full_extended_uml.txt)
![uml-ext-geopf](https://github.com/user-attachments/assets/2c156123-d723-4aa7-8f42-33a404d574be)


Mise à jour plus complète sur les classes et widgets

**Reply by lowzonenose** [reply]
Pour info,
Theme, Legend, Layers... sont des classes appelées par **Editor**.
Seul **Editor** est un classe de type **Control**.

La classe **Control** peut aussi gérer les _properties_ non natives de OpenLayers, ex. position (la fameuse position relative) ou alors une classe Interface pour la position des contrôles...

**Comment by IGNFhc** [comment]
Bonjour aujourd'hui nous avons décidé d'orienter les prochaines étapes de travail du "DOM core" vers :
- lister des "MapComponents" à développer (par exemple : des panels, des boutons, inputs de coordonées, ....) avec leurs entrées / sorties / méthodes / évenements 
- trouver un système de templating satisfaisant 
- Réfléchir à la gestion du css (sans doute conserver le fonctionnement actuel à 3fichiers)
- proposition de créer une classe CanvasOMManager pour gérer : les collisions, ouverture, fermeture de panel / infobulles, superposition, placement sur le canvas selon espace dispo (quadrillage, ...), type de dom, .....

Prochaine date : mi juin

**Comment by azarz** [comment]
Pour un seul dev utilisable dans plusieurs librairies carto, concept d'adapter à creuser comme sur https://github.com/JamesLMilner/terra-draw

**Reply by lowzonenose** [reply]
yeah, il va falloir apprendre nos *design pattern* !

**Comment by elias75015** [comment]
https://github.com/JamesLMilner/terra-draw

**Comment by elias75015** [comment]
CR rencontre CdB

Fork du projet pour le projet de cartographie des services publics
-> https://github.com/cdebarros/geopf-extensions-openlayers
Page de démo :
-> https://cartosp.modernisation.gouv.fr/



Copiés - collés pour créer le nouveaux widgets.

Gain de temps pour gestion du DSFR et certains widgets clés en main (rechercher par adresse, layerswitcher)
Un template de développement plus visible aurait été utile.
Le Contributing.md est à revoir et mettre à jour.

A court terme : accueiilir les contributions dans un dossier contributions. Ajouter la possibilité de compiler le projet avec ou sans ses contributions. Encore mieux, pouvoir choisir les contributions que l'on veut compiler.

Isochrone surchargé : voir si on peut récupérer les évplutions dans le widget principal.



Attention, les contributions peuvent être au mode DSFR, Classic, ou les deux. Il faudrait trouver un moyen de les documenter (ou de les ranger) correctement pour que ce soit explicite pour l'utilisateur.

**Comment by lowzonenose** [comment]
Un concept que j'ai mis en place sur le widget Reporting à creuser pour notre refactoring des API
[Ioc.pdf](https://github.com/user-attachments/files/20942097/Ioc.pdf)

**Reply by elias75015** [reply]
@IGNFhc :eyes:

**Comment by IGNFhc** [comment]
oui j'ai lu le pdf merci JP c'est très instructif

**Comment by MatRouillard** [comment]
Il pourrait être intéressant de penser les contrôles de manière modulaire et "généralisable", pour créer plus facilement des contrôles sur la carte. Notamment en créant par exemple un contrôle général qui construirait :

- un bouton sur la carte, positionné selon ce qui est possible de faire avec les contrôles (en rajoutant peut-être aussi une option pour le mettre aux positions top, left, bottom et right, et pas restreindre au coin ?)
- Des options pour le bouton, permettant de choisir facilement une icône (parmi celles du dsfr / remixicon ou autre aussi si possible ?) et un titre au survol
- une option permettant de définir si le contrôle fonctionne avec une modale, avec seulement une interaction (type zoom) ou les deux (comme l'édition)

Si c'est un contrôle qui ouvre une modale:

- une modale avec juste un bouton fermer (à minima la croix), et potentiellement des options de style de modales (du genre : modale comme le LayerSwitcher, avec une panel violet, peu de padding etc. / modale dsfr classique pour les formulaires)
- optionnellement un titre (avec différentes tailles peut-être ?)
- Un body préconfiguré, qui peut être soit instancié via le paramètre correspondant à la création du contrôle, ou bien via un setter plus tard (style setDialogElement ou quelque chose comme ça)

Si c'est un contrôle avec une interaction:
- pouvoir passer là ou les interactions que ça met, avec notamment une fonction de callback qui permet de lier map et interactions

Si c'est les deux, alors lier les deux

Ça pourrait être intéressant aussi de lier d'autres contrôles peut-être (au choix de l'utilisateur) style le LayerSwitcher ou si c'est un contrôle agissant sur une couche et pas sur la carte, lié une couche au contrôle avec des setter et getter accessibles publiquement.
 
De la même manière que pour le contrôle, des fonctions utilitaires permettant notamment de créer des élements de formulaire (type input, select, checkbox, toggle et slider, ou du texte) possédant déjà les bonnes classes DSFR / classique et surtout tous les attributs aria nécessaire serait pratique et permettrait de rentrer dans cette optique de contrôle généralistes.

Cela ne concerne peut-être pas tous les contrôles, donc à voir comment cela pourrait être mis en place.

**Reply by elias75015** [reply]
@IGNFhc :eyes: à ton retour
